### PR TITLE
[JAX] Added unit tests for distributed LayernormMLP

### PR DESCRIPTION
**Pull request by phu0ngng**
This PR added unit tests for distributed `fused_layernorm_fp8_mlp` and `LayerNormMLP`.
The outputs of the run on multiple GPUs are compared again to the one with a single GPU for correctness check.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/jax/test_distributed_layernorm_mlp.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+            [4, (2, 2), ('fsdp', 'tp'),
+             MeshResource(fsdp_resource='fsdp', tp_resource='tp')])
+    return configs
+
+
+class TestDistributedLayernormMLP:
+
+    def generate_inputs(self, input_shape, activation_type, use_bias, dtype):
+        # m, n, k = shape
+        batch, seqlen, hidden_in = input_shape
```
Comment (zlsh80826) [inline tests/jax/test_distributed_layernorm_mlp.py:55]: Here is not aligned to the comment of line 29. Here is b, s, h, but the comment of line 29 is s, b, h.

File: tests/jax/test_distributed_layernorm_mlp.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+from transformer_engine.jax.sharding import HIDDEN_AXES, HIDDEN_TP_AXES, \
+    BATCH_AXES, SEQLEN_TP_AXES, SEQLEN_AXES, \
+    W_NO_SHARD_AXES, W_FSDP_AXES, W_TP_AXES, W_JOINED_AXES
+from transformer_engine.jax.sharding import MeshResource
+from utils import assert_allclose, is_devices_enough
+
+
+is_fp8_supported, reason = is_fp8_available()
+DTYPES = [jnp.bfloat16, jnp.float16]
+INPUT_SHAPE = [[64, 64, 32]]    # [seqlen, batch, hidden_in]
```
Comment (zlsh80826) [inline tests/jax/test_distributed_layernorm_mlp.py:29]: Let's use a shape with batch != seqlen, which can increase the coverage. Otherwise, we may miss some corner case for something related to`transpose_bs` .

File: tests/jax/test_distributed_layernorm_mlp.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+                          layernorm_type: str = "rmsnorm",
+                          activation_type: Sequence[Union[str, Callable]] = ('gelu',),
+                          use_bias: bool = True,
+                          multi_gpus: bool = False,
+                          ) -> jnp.ndarray:
+
+        fp8_meta_pkg = FP8MetaPackage(2, fp8_max, fp8_metas_amax, fp8_metas_scale,
+                                      fp8_metas_scale_inv)
+
+        if (multi_gpus):
```
Comment (zlsh80826) [inline tests/jax/test_distributed_layernorm_mlp.py:91]: Nitpick: the parentheses is not needed.

File: tests/jax/test_distributed_layernorm_mlp.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+                out_shardings=out_shardings,
+                static_argnums=range(len(multi_inputs),
+                                     len(static_inputs)+len(multi_inputs)+1))   # +1 for multi_gpus
+
+            multi_fwd, multi_grads = multi_jitter(*multi_inputs, *static_inputs, True)
+
+        assert_allclose(multi_fwd, single_fwd, dtype=dtype)
+        for i in range(len(inputs)):
+            if multi_grads[i] is not None:
+                assert_allclose(multi_grads[i], single_grads[i], dtype=dtype)
```
Comment (zlsh80826) [inline tests/jax/test_distributed_layernorm_mlp.py:184]: When wrapping a `assert_xxx` into a for loop, it is important to include an error message. so that we can quick identify which tensors have issue. This helps us quickly identify which tensors might have issues. Could you add an error message to the assert_allclose arguments? For example, something like err_msg=f'{grads[i]} is not close'.

File: tests/jax/test_distributed_layernorm_mlp.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                dot_2_input_axes=DOT_2_INPUT_AXES,
+                name='mlp'
+            )
+            params_sharded = ln_mlp_sharded.init(init_rngs, x)
+            mlp_out_sharded, ln_out_sharded = ln_mlp_sharded.apply(params_sharded, x,
+                                                                   deterministic=True)
+
+        # Make sure params values are the same
+        jax.tree_util.tree_map(lambda x, y: assert_allclose(x, y, dtype=dtype), \
+                               params_sharded['params'], params_single['params'])
```
Comment (zlsh80826) [inline tests/jax/test_distributed_layernorm_mlp.py:239]: There is a `assert_tree_like_allclose` under utils.py

**Comment by phu0ngng** [comment]
\te-ci jax

**Review by zlsh80826 — APPROVED** [review]
LGTM!
